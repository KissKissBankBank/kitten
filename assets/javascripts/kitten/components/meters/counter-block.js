import React, { Component, Fragment } from 'react'
import Radium, { StyleRoot } from 'radium'
import { Title as TitleBase } from 'kitten/components/typography/title'
import { Text as TextBase } from 'kitten/components/typography/text'
import { Marger } from 'kitten/components/layout/marger'
import { ScreenConfig } from 'kitten/constants/screen-config'
import { withMediaQueries } from 'kitten/hoc/media-queries'

const Text = Radium(TextBase)
const Title = Radium(TitleBase)

class CounterBlockBase extends Component {
  render() {
    const {
      title,
      titleTag,
      subTitle,
      subTitleTag,
      counter,
      viewportIsSOrLess,
      titleProps,
    } = this.props
    const titleSize = viewportIsSOrLess ? 'senary' : 'tertiary'
    const margin = viewportIsSOrLess ? 2 : 4

    return (
      <StyleRoot style={styles.counterBlock}>
        <Marger bottom={margin}>
          <Title modifier={titleSize} {...titleProps}>
            {title}
          </Title>
        </Marger>
        <Text
          style={styles.counterText}
          fontStyle="italic"
          weight="bold"
          tag={titleTag}
        >
          {counter}
        </Text>
        <Marger top={margin}>
          <Title
            style={styles.subTitle}
            modifier="senary"
            margin={false}
            tag={subTitleTag}
            italic
          >
            {subTitle}
          </Title>
        </Marger>
      </StyleRoot>
    )
  }
}

CounterBlockBase.defaultProps = {
  title: null,
  titleTag: 'p',
  subTitle: null,
  subTitleTag: 'p',
  counter: '0',
}

const styles = {
  counterBlock: {
    textAlign: 'center',
  },

  counterText: {
    fontSize: '4.125rem',
    lineHeight: 0.7,
    whiteSpace: 'nowrap',
    display: 'block',
    paddingBottom: 4,
    margin: 0,

    [`@media (min-width: ${ScreenConfig.M.min}px)`]: {
      fontSize: '8.25rem',
      paddingBottom: 7,
    },

    [`@media (min-width: ${ScreenConfig.L.min}px)`]: {
      fontSize: '9.9rem',
      paddingBottom: 10,
    },
  },

  subTitle: {
    [`@media (max-width: ${ScreenConfig.S.max}px)`]: {
      fontSize: '.875rem',
    },
  },
}

export const CounterBlock = withMediaQueries({
  viewportIsSOrLess: true,
})(CounterBlockBase)
