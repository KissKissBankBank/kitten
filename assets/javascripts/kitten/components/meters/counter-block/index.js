import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import { Title } from '../../../components/typography/title'
import { Text } from '../../../components/typography/text'
import { Marger } from '../../../components/layout/marger'
import { HorizontalStroke } from '../../../components/layout/horizontal-stroke'
import { ScreenConfig } from '../../../constants/screen-config'
import { pxToRem } from '../../../helpers/utils/typography'

const Container = styled.div`
  text-align: center;

  .styledText {
    line-height: 0.7;
    white-space: nowrap;
    display: flex;
    padding-bottom: ${pxToRem(4)};
    margin: 0;
    flex-grow: 1;
    font-size: ${pxToRem(50)};
    justify-content: center;

    @media (min-width: ${ScreenConfig.XS.min}px) {
      font-size: ${pxToRem(66)};
    }

    @media (min-width: ${ScreenConfig.S.min}px) {
      font-size: ${pxToRem(96)};
      padding-bottom: ${pxToRem(6)};
    }

    @media (min-width: ${ScreenConfig.M.min}px) {
      font-size: ${pxToRem(116)};
      padding-bottom: ${pxToRem(7)};
    }

    @media (min-width: ${ScreenConfig.L.min}px) {
      font-size: ${pxToRem(158)};
      padding-bottom: ${pxToRem(10)};
    }
  }

  .styledSubTitle {
    margin: 0;
  }
`

const FlexContainer = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
`

const StrokeContainer = styled.div`
  flex-grow: 2;

  .styledStroke {
    width: 100%;
    height: ${pxToRem(6)};
  }
`

const Unit = styled.span`
  display: inline-flex;
  align-self: center;
  font-size: 0.5em;
`

const isString = val => typeof val === 'string'

export const CounterBlock = ({
  title,
  titleTag,
  subTitle,
  subTitleTag,
  counter,
  unit,
}) => {
  const margin = { default: 2, fromS: 4 }

  return (
    <Container>
      {isString(title) ? (
        <Marger bottom={margin}>
          <Title modifier="tertiary" noMargin tag={titleTag}>
            {title}
          </Title>
        </Marger>
      ) : (
        title
      )}
      <FlexContainer>
        <StrokeContainer className="k-u-hidden@m-down">
          <HorizontalStroke className="styledStroke" size="big" />
        </StrokeContainer>
        {isString(counter) ? (
          <Text
            className="styledText"
            weight="bold"
            tag={titleTag}
            fontStyle="italic"
          >
            {counter}
            {unit && <Unit>&nbsp;{unit}</Unit>}
          </Text>
        ) : (
          counter
        )}

        <StrokeContainer className="k-u-hidden@m-down">
          <HorizontalStroke className="styledStroke" size="big" />
        </StrokeContainer>
      </FlexContainer>
      {isString(subTitle) ? (
        <Marger top={margin}>
          <Text
            className="styledSubTitle"
            size="huge"
            tag={subTitleTag}
            tag="p"
            fontStyle="italic"
          >
            {subTitle}
          </Text>
        </Marger>
      ) : (
        subTitle
      )}
    </Container>
  )
}

CounterBlock.propTypes = {
  title: PropTypes.node,
  titleTag: PropTypes.string,
  subTitle: PropTypes.node,
  subTitleTag: PropTypes.string,
  counter: PropTypes.node,
  unit: PropTypes.string,
}

CounterBlock.defaultProps = {
  title: null,
  titleTag: 'p',
  subTitle: null,
  subTitleTag: 'p',
  counter: '0',
  unit: null,
}
