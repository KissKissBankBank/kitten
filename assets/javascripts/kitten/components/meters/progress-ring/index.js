import React from 'react'
import PropTypes from 'prop-types'
import styled, { keyframes } from 'styled-components'
import COLORS from '../../../constants/colors-config'
import { pxToRem } from '../../../helpers/utils/typography'

const getDashLength = radius => 2 * Math.PI * radius

const getDashOffset = ({ r, progressValue }) => (
  getDashLength(r) * (100 - progressValue) / 100
)

const rotateAnimate = ({ r, progressValue }) => keyframes`
  from {
    stroke-dashoffset: ${getDashLength(r)};
  }
  to {
    stroke-dashoffset: ${getDashOffset({ r, progressValue })};
  }
`

const StyledCircleBackground = styled.circle`
  fill: transparent;
  stroke: ${COLORS.line1};
  stroke-width: ${({ strokeWidth }) => pxToRem(strokeWidth)};
`

const StyledCircle = styled.circle`
  fill: transparent;
  stroke-width: ${({ strokeWidth }) => pxToRem(strokeWidth)};
  stroke-linecap: butt;
  stroke-dasharray: ${({ r }) => getDashLength(r)};
  stroke-dashoffset: ${({ r, progressValue }) =>
    getDashOffset({ r, progressValue })};
  stroke: ${({ progressColor }) => progressColor};
  transform: rotate(-90deg);
  transform-origin: 50% 50%;
  animation: ${rotateAnimate} 1.4s ease-out;
`

export const ProgressRing = ({
  color,
  value,
  width,
  strokeWidth,
  ...others
}) => {
  const circleX = width / 2
  const circleY = width / 2

  const radius = circleX - strokeWidth

  const viewBox = `0 0 ${width} ${width}`

  const progressValue = value < 100 ? value : 100

  return (
    <svg
      {...others}
      width={width}
      height={width}
      viewBox={viewBox}
      xmlns="http://www.w3.org/2000/svg"
    >
      <StyledCircleBackground
        cx={circleX}
        cy={circleY}
        r={radius}
        strokeWidth={strokeWidth}
      />
      <StyledCircle
        progressColor={color}
        progressValue={progressValue}
        cx={circleX}
        cy={circleY}
        r={radius}
        strokeWidth={strokeWidth}
      />
    </svg>
  )
}

ProgressRing.defaultProps = {
  color: COLORS.primary1,
  value: '',
  width: 50,
  radius: null,
  strokeWidth: 5,
}

ProgressRing.propTypes = {
  /**
    ProgressRing color
  */
  color: PropTypes.string,
  /**
    Percentage of progress %
  */
  value: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  /**
    Width of ProgressRing
  */
  width: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  /**
    Radius value of ProgressRing
  */
  radius: PropTypes.number,
   /**
    Width of circle (stroke)
  */
  strokeWidth: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
}
