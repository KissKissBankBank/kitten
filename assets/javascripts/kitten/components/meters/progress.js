import React, { Component } from 'react'
import classNames from 'classnames'

export class Progress extends Component {
  render() {
    const {
      className,
      color,
      value,
      rampProps,
      ...others,
    } = this.props

    const progressClassName = classNames(
      'k-Progress',
      className,
    )

    const style = {
      backgroundColor: `${color}`,
      width: `${value}%`,
    }

    const rampClassName = classNames(
      'k-Progress__ramp',
      rampProps.className,
    )

    return (
      <div
        { ...others }
        role="progressbar"
        aria-valuemin="0"
        aria-valuemax="100"
        aria-valuenow={ this.props.value }
        className={ progressClassName }>
        <div { ...rampProps } className={ rampClassName }>
          <div
            className="k-Progress__slider"
            style={ style }
          />
        </div>
      </div>
    )
  }
}

Progress.defaultProps = {
  className: null,
  color: null,
  value: 50,
  rampProps: {},
}
