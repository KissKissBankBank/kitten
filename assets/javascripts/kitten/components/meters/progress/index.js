import React, { useEffect, useState } from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import classNames from 'classnames'
import COLORS from '../../../constants/colors-config'
import { pxToRem } from '../../../helpers/utils/typography'

const valueMin = 0
const valueMax = 100

const StyledProgress = styled.div`
  max-width: 100%;
`

const StyledRamp = styled(({ style, sliderColor, progressValue, ...props }) => (
  <div {...props} />
))`
  position: relative;
  height: ${pxToRem(2)};
  background: ${COLORS.line1};
  ${({ style }) => style}

  &::after {
    content: '';
    position: absolute;
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;
    max-width: ${({ progressValue }) => progressValue};
    transition: max-width 1s cubic-bezier(0, 0.5, 0.3, 1);
    background: ${({ sliderColor }) => sliderColor};

    ${StyledProgress}.is-disabled & {
      background: ${COLORS.line2};
    }
  }
`

export const Progress = ({ color, value, rampProps, ...others }) => {
  const [progressValue, setProgressValue] = useState(0)

  useEffect(() => {
    let progress = 0

    if (value < valueMin) progress = valueMin
    else if (value > valueMax) progress = valueMax
    else progress = value

    setProgressValue(progress)
  }, [value])

  return (
    <StyledProgress
      {...others}
      role="progressbar"
      aria-valuemin={valueMin}
      aria-valuemax={valueMax}
      aria-valuenow={progressValue}
    >
      <StyledRamp
        {...rampProps}
        sliderColor={color}
        progressValue={`${progressValue}%`}
      />
    </StyledProgress>
  )
}

Progress.defaultProps = {
  color: COLORS.primary1,
  value: 50,
  rampProps: {},
}

Progress.propTypes = {
  color: PropTypes.string,
  value: PropTypes.number,
  rampProps: PropTypes.object,
}
