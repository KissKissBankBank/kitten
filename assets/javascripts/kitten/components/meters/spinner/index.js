import React, { useEffect, useState } from 'react'
import PropTypes from 'prop-types'
import styled, { keyframes } from 'styled-components'
import classNames from 'classnames'
import COLORS from '../../../constants/colors-config'
import { pxToRem } from '../../../helpers/utils/typography'

const circleRadius = 30
const dashLength = `calc(2 * 3.141592 * ${pxToRem(circleRadius)})`

const rotateAnimate = keyframes`
  from {
    stroke-dashoffset: ${dashLength};
  }
  to {
    stroke-dashoffset: calc(${dashLength} * ${({ progressValue }) =>
  100 - progressValue} / 100);
  }
`

const StyledCircle = styled(({ progressColor, progressValue, ...props }) => (
  <circle {...props} />
))`
  fill: transparent;
  stroke-width: ${pxToRem(4)};
  stroke-linecap: square;
  stroke-dasharray: ${dashLength};
  stroke-dashoffset: calc(
    ${dashLength} * ${({ progressValue }) => 100 - progressValue} / 100
  );
  stroke: ${({ progressColor }) => progressColor};

  transform: rotate(-90deg);
  transform-origin: 50% 50%;
  animation: ${rotateAnimate} 3s ease-out;
`

const StyledCircleBackground = styled.circle`
  fill: transparent;
  stroke: ${COLORS.line1};
  stroke-width: ${pxToRem(4)};
`

export const SpinnerProgress = ({ color, value, rampProps, ...others }) => (
  <svg
    {...rampProps}
    width="70"
    height="70"
    viewBox="0 0 70 70"
    xmlns="http://www.w3.org/2000/svg"
  >
    <StyledCircleBackground cx="35" cy="35" r={circleRadius} />
    <StyledCircle
      progressColor={color}
      progressValue={value}
      cx="35"
      cy="35"
      r={circleRadius}
    />
  </svg>
)

SpinnerProgress.defaultProps = {
  color: COLORS.primary1,
  value: '',
  rampProps: {},
}

SpinnerProgress.PropTypes = {
  color: PropTypes.string,
  value: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  rampProps: PropTypes.object,
}
