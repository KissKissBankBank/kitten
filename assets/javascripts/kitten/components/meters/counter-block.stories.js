import React from 'react'
import { storiesOf } from '@storybook/react'
import { withInfo } from '@storybook/addon-info'
import { withKnobs, text } from '@storybook/addon-knobs/react'
import { CounterBlock } from '../../components/meters/counter-block'
import { Grid, GridCol } from '../../components/grid/grid'
import { Marger } from '../../components/layout/marger'

const createMockMediaMatcher = matches => () => ({
  matches,
  addListener: () => {},
  removeListener: () => {},
})

storiesOf('metrics/CounterBlock', module)
  .addDecorator(withKnobs)
  .add(
    'default',
    withInfo('common info')(() => (
      <Marger top="3" bottom="3">
        <Grid>
          <GridCol offset-l="1" col-l="10">
            <CounterBlock
              title={text('Title', 'Total validated amount')}
              subTitle={text(
                'Sub Title',
                'euro collected since Novembre 19th 2014',
              )}
              counter={text('Counter', '13 106 700')}
            />
          </GridCol>
        </Grid>
      </Marger>
    )),
  )
