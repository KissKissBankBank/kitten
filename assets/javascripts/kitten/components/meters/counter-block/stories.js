import React from 'react'
import { storiesOf } from '@storybook/react'
import { withInfo } from '@storybook/addon-info'
import { withKnobs, text } from '@storybook/addon-knobs'
import { CounterBlock } from '../../../components/meters/counter-block'
import { Marger } from '../../../components/layout/marger'

storiesOf('Meters/CounterBlock', module)
  .addDecorator(withKnobs)
  .add(
    'default',
    withInfo('Common info')(() => (
      <Marger top="3" bottom="3">
        <CounterBlock
          title={text('Title', 'Total validated amount')}
          subTitle={text('Sub Title', 'collected since Novembre 19th 2014')}
          counter={text('Counter', '52 982 287')}
          unit={text('Unit', 'â‚¬')}
        />
      </Marger>
    )),
  )
