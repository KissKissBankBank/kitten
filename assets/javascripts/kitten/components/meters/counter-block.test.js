import React from 'react'
import renderer from 'react-test-renderer'
import { CounterBlock } from 'kitten/components/meters/counter-block'

const createMockMediaMatcher = matches => () => ({
  matches,
  addListener: () => {},
  removeListener: () => {},
})

describe('<CounterBlock />', () => {
  let originalMatchMedia
  let component

  beforeEach(() => {
    originalMatchMedia = window.matchMedia
  })

  afterEach(() => {
    window.matchMedia = originalMatchMedia
  })

  describe('with default props', () => {
    beforeEach(() => {
      window.matchMedia = createMockMediaMatcher(false)
      component = renderer
        .create(<CounterBlock title="foo" subTitle="bar" counter="13 106" />)
        .toJSON()
    })

    it('matches with snapshot', () => {
      expect(component).toMatchSnapshot()
    })
  })
})
