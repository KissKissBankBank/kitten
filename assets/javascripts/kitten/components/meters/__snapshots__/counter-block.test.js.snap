// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CounterBlock /> with default props matches with snapshot 1`] = `
<div
  data-radium={true}
  style={
    Object {
      "textAlign": "center",
    }
  }
>
  <div
    className="marger__StyledMarger-q3lecu-0 iYDxXp"
  >
    <p
      className="k-Title k-Title--tertiary k-Title--withoutMargin"
    >
      foo
    </p>
  </div>
  <div
    data-radium={true}
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "justifyContent": "center",
      }
    }
  >
    <div
      className="k-u-hidden@m-down"
      data-radium={true}
      style={
        Object {
          "flexGrow": 2,
        }
      }
    >
      <span
        className="horizontal-stroke__StyledHorizontalStroke-sc-15k7x5m-0 kGvLel"
        height={null}
        size="big"
        width={null}
      />
    </div>
    <p
      className="rmq-2bf6e1a2 rmq-ca40c6f7 rmq-1103da81 rmq-650fa8a3 k-u-style-italic k-u-weight-bold"
      data-radium={true}
      style={
        Object {
          "display": "block",
          "flexGrow": 1,
          "fontSize": "3rem",
          "lineHeight": 0.7,
          "margin": 0,
          "paddingBottom": 4,
          "textIndent": "-1.5rem",
          "whiteSpace": "nowrap",
        }
      }
    >
      13 106
    </p>
    <div
      className="k-u-hidden@m-down"
      data-radium={true}
      style={
        Object {
          "flexGrow": 2,
        }
      }
    >
      <span
        className="horizontal-stroke__StyledHorizontalStroke-sc-15k7x5m-0 kGvLel"
        height={null}
        size="big"
        width={null}
      />
    </div>
  </div>
  <div
    className="marger__StyledMarger-q3lecu-0 cXVUrc"
  >
    <p
      className="k-u-size-huge k-u-style-italic k-u-weight-bold"
      data-radium={true}
      style={
        Object {
          "margin": 0,
        }
      }
    >
      bar
    </p>
  </div>
  <style
    dangerouslySetInnerHTML={
      Object {
        "__html": "@media (min-width: 480px){ .rmq-2bf6e1a2{font-size: 4.125rem !important;}}
@media (min-width: 640px){ .rmq-ca40c6f7{font-size: 6rem !important;
padding-bottom: 6px !important;}}
@media (min-width: 768px){ .rmq-1103da81{font-size: 7.3rem !important;
padding-bottom: 7px !important;}}
@media (min-width: 1080px){ .rmq-650fa8a3{font-size: 9.9rem !important;
padding-bottom: 10px !important;}}",
      }
    }
  />
</div>
`;
